batch2/batch/batch.py

```diff
@@ -1,19 +1,1278 @@
+@prom_async_time(REQUEST_TIME_POST_CREATE_BATCH)
+@rest_authenticated_users_only
+async def create_batch(request, userdata):
+    start = time.time()
+    parameters = await request.json()
+
+    validator = cerberus.Validator(schemas.batch_schema)
+    if not validator.validate(parameters):
+        abort(400, 'invalid request: {}'.format(validator.errors))
+
+    batch = await Batch.create_batch(
+        attributes=parameters.get('attributes'),
+        callback=parameters.get('callback'),
+        userdata=userdata,
+        n_jobs=parameters['n_jobs'])
+    if batch is None:
+        abort(400, f'creation of batch in db failed')
+
+    log.info(f'took {round(time.time() - start, 3)} seconds to initialize batch {batch.id} in db')
+    return jsonify(await batch.to_dict(include_jobs=False))
+
+
+async def _get_batch(batch_id, user, limit=None, offset=None):
+    batch = await Batch.from_db(batch_id, user)
+    if not batch:
+        abort(404)
+    return await batch.to_dict(include_jobs=True, limit=limit, offset=offset)
+
+
+async def _cancel_batch(batch_id, user):
+    batch = await Batch.from_db(batch_id, user)
+    if not batch:
+        abort(404)
+    asyncio.ensure_future(batch.cancel())
+
+
+async def _get_job(batch_id, job_id, user):
+    job = await Job.from_db(batch_id, job_id, user)
+    if not job:
+        abort(404)
+    return job.to_dict()
+
+
+@routes.get('/api/v1alpha/batches/{batch_id}')
+@prom_async_time(REQUEST_TIME_POST_GET_BATCH)
+@rest_authenticated_users_only
+async def get_batch(request, userdata):
+    batch_id = int(request.match_info['batch_id'])
+    user = userdata['username']
+    params = request.query
+    limit = params.get('limit')
+    offset = params.get('offset')
+    return jsonify(await _get_batch(batch_id, user, limit=limit, offset=offset))
+
+
+@routes.patch('/api/v1alpha/batches/{batch_id}/cancel')
+@prom_async_time(REQUEST_TIME_PATCH_CANCEL_BATCH)
+@rest_authenticated_users_only
+async def cancel_batch(request, userdata):
+    batch_id = int(request.match_info['batch_id'])
+    user = userdata['username']
+    await _cancel_batch(batch_id, user)
+    return jsonify({})
+
+
+@routes.patch('/api/v1alpha/batches/{batch_id}/close')
+@prom_async_time(REQUEST_TIME_PATCH_CLOSE_BATCH)
+@rest_authenticated_users_only
+async def close_batch(request, userdata):
+    batch_id = int(request.match_info['batch_id'])
+    user = userdata['username']
+    batch = await Batch.from_db(batch_id, user)
+    if not batch:
+        abort(404)
+    await batch.close()
+    return jsonify({})
+
+
+@routes.delete('/api/v1alpha/batches/{batch_id}')
+@prom_async_time(REQUEST_TIME_DELETE_BATCH)
+@rest_authenticated_users_only
+async def delete_batch(request, userdata):
+    batch_id = int(request.match_info['batch_id'])
+    user = userdata['username']
+    batch = await Batch.from_db(batch_id, user)
+    if not batch:
+        abort(404)
+    asyncio.ensure_future(batch.mark_deleted())
```

Comment:
mark_deleted doesn't do much work besides call cancel which itself uses ensure_future, so I don't think we should use ensure future here.